File: combinatorics/resource/ST/project_5/bai_toan_4/trees/graph/graph_to_fcns.py
def build_fcns(u, p, adj, fc, ns):
    children = [v for v in adj[u] if v != p]
    if children:
        fc[u] = children[0]
        for i in range(len(children) - 1):
            ns[children[i]] = children[i + 1]
    for v in children:
        build_fcns(v, u, adj, fc, ns)

n = int(input("Nhap so dinh: "))
adj = [[] for _ in range(n)]

print("Nhap danh sach ke:")
for i in range(n):
    line = list(map(int, input(f"{i}: ").split()))
    adj[i].extend(line)

fc = [-1] * n
ns = [-1] * n

build_fcns(0, -1, adj, fc, ns)

print("Mang fc:", fc)
print("Mang ns:", ns)

File: combinatorics/resource/ST/project_5/bai_toan_4/trees/graph/graph_to_fcns.cpp
#include <iostream>
#include <vector>
using namespace std;

void buildFCNS(int u, int p, const vector<vector<int>>& adj,
               vector<int>& fc, vector<int>& ns) {
    vector<int> children;
    for (int v : adj[u]) {
        if (v != p)
            children.push_back(v);
    }

    if (!children.empty()) {
        fc[u] = children[0];
        for (size_t i = 0; i + 1 < children.size(); ++i)
            ns[children[i]] = children[i + 1];
    }

    for (int v : children)
        buildFCNS(v, u, adj, fc, ns);
}

int main() {
    int n;
    cout << "Nhap so dinh: ";
    cin >> n;
    cin.ignore();

    vector<vector<int>> adj(n);
    cout << "Nhap danh sach ke:\n";
    for (int i = 0; i < n; ++i) {
        cout << i << ": ";
        string line;
        getline(cin, line);
        stringstream ss(line);
        int v;
        while (ss >> v)
            adj[i].push_back(v);
    }

    vector<int> fc(n, -1), ns(n, -1);
    buildFCNS(0, -1, adj, fc, ns);

    cout << "Mang fc: ";
    for (int x : fc) cout << x << " ";
    cout << "\nMang ns: ";
    for (int x : ns) cout << x << " ";
    cout << "\n";

    return 0;
}